cmd/openshift-install/gather.go

```diff
@@ -0,0 +1,231 @@
+)
+
+func newGatherCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "gather",
+		Short: "Gather debugging data for a given installation failure",
+		Long: `Gather debugging data for a given installation failure.
+
+When installation for Openshift cluster fails, gathering all the data useful for debugging can
+become a difficult task. This command helps users to collect the most relevant information that can be used
+to debug the installation failures`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return cmd.Help()
+		},
+	}
+	cmd.AddCommand(newGatherBootstrapCmd())
+	return cmd
+}
+
+var (
+	gatherBootstrapOpts struct {
+		bootstrap string
+		masters   []string
+	}
+)
+
+func newGatherBootstrapCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "bootstrap",
+		Short: "Gather debugging data for a failing to bootstrap control plane",
+		Args:  cobra.ExactArgs(0),
+		RunE:  runGatherBootstrapCmd,
+	}
+	cmd.PersistentFlags().StringVar(&gatherBootstrapOpts.bootstrap, "bootstrap", "", "")
+	cmd.PersistentFlags().StringArrayVar(&gatherBootstrapOpts.masters, "master", []string{}, "")
+	return cmd
+}
+
+func runGatherBootstrapCmd(_ *cobra.Command, _ []string) error {
+	cleanup := setupFileHook(rootOpts.dir)
+	defer cleanup()
+
+	tfStateFilePath := filepath.Join(rootOpts.dir, terraform.StateFileName)
+	_, err := os.Stat(tfStateFilePath)
+	if os.IsNotExist(err) {
+		return unSupportedPlatformGather()
+	}
+	if err != nil {
+		return err
+	}
+
+	assetStore, err := assetstore.NewStore(rootOpts.dir)
+	if err != nil {
+		return errors.Wrap(err, "failed to create asset store")
+	}
+
+	config := &installconfig.InstallConfig{}
+	if err := assetStore.Fetch(config); err != nil {
+		return errors.Wrapf(err, "failed to fetch %s", config.Name())
+	}
+
+	sfRaw, err := ioutil.ReadFile(tfStateFilePath)
+	if err != nil {
+		return errors.Wrapf(err, "failed to read %q", tfStateFilePath)
+	}
+
+	var tfstate terraformState
+	if err := json.Unmarshal(sfRaw, &tfstate); err != nil {
+		return errors.Wrapf(err, "failed to unmarshal %q", tfStateFilePath)
+	}
+
+	bootstrap, masters, err := extractHostAddresses(config.Config, tfstate)
+	if err != nil {
+		if err2, ok := err.(errUnSupportedGatherPlatform); ok {
+			logrus.Error(err2)
+			return unSupportedPlatformGather()
+		}
+		return errors.Wrapf(err, "failed to get bootstrap and control plane host addresses from %q", tfStateFilePath)
+	}
+
+	logrus.Info("Use the following commands to gather logs from the cluster")
+	logrus.Infof("> ssh core@%s -A '/usr/local/bin/installer-gather.sh %s'", bootstrap, strings.Join(masters, " "))
+	logrus.Infof("> scp core@%s:~/installer-logs.tar.gz .", bootstrap)
+
+	return nil
+}
+
+func logGatherBootstrap(boostrap string, masters []string) {
+	if s, ok := os.LookupEnv("SSH_AUTH_SOCK"); !ok || s == "" {
+		logrus.Info("Make sure ssh-agent is running, env SSH_AUTH_SOCK is set to the ssh-agent's UNIX socket and your private key is added to the agent.")
```

Comment:
How portable is `ssh-agent` and `SSH_AUTH_SOCK`?  My impression is that this will be POSIX- (or maybe even Linux-)specific.

Reply:
the goal is to make sure the user has agent running, this is an easy check to report user that it might not.

`ssh-agent` works on both linux and macOS using the same env. and we don't do windows yet ;)